Sources/Prelude/Refinements.swift

```diff
@@ -0,0 +1,136 @@
+
+extension Refined: Hashable where A: Hashable { }
+
+public extension Refinement {
+    /// convenient way initialize a value of type `Refined<RefinedType, Self>` given a proposed value. This is equivalent to calling `try? Refined<RefinedType, Self>.init(value)`, but much easier to type
+    ///
+    /// - Parameter value: the proposed value
+    /// - Returns: a value of type `Refined<RefinedType, Self>`, or nil
+    static func of(_ value: RefinedType) -> Refined<RefinedType, Self>? {
+        return try? .init(value)
+    }
+}
+
+// MARK: - Both
+
+/// a refinement that can be used to combine any two other refinements (`L` and `R`), such that they both must pass
+public enum Both<L: Refinement, R: Refinement>: Refinement where L.RefinedType == R.RefinedType {
+    public typealias RefinedType = L.RefinedType
+    public static func isValid(_ value: RefinedType) -> Bool {
+        return L.isValid(value) && R.isValid(value)
+    }
+}
+
+/// given a value that is refined by two rules (`L` and `R`), produce a value that is refined only by the rule on the left (`L`). This function will always succeed (the underlying value has already passed both rules)
+///
+/// - Parameter refined: a value that is refined by two rules
+/// - Returns: the same value, refined only by the rule on the left
+public func left<A, L, R>(_ refined: Refined<A, Both<L, R>>) -> Refined<A, L> {
+    guard let result = L.of(refined.value) else {
+        preconditionFailure("when calling `left(_:)`: the value “\(refined.value)” didn’t satisfy the refinement “\(L.self)”. This shouldn’t happen. Was the value mutated after refinements had already been checked?")
+    }
+    return result
+}
+
+/// given a value that is refined by two rules (`L` and `R`), produce a value that is refined only by the rule on the right (`R`). This function will always succeed (the underlying value has already passed both rules)
+///
+/// - Parameter refined: a value that is refined by two rules
+/// - Returns: the same value, refined only by the rule on the right
+public func right<A, L, R>(_ refined: Refined<A, Both<L, R>>) -> Refined<A, R> {
+    guard let result = R.of(refined.value) else {
+        preconditionFailure("when calling `right(_:)`: the value “\(refined.value)” didn’t satisfy the refinement “\(R.self)”. This shouldn’t happen. Was the value mutated after refinements had already been checked?")
+    }
+    return result
+}
+
+// MARK: - Not
+
+/// a refinement that can be used to invert any other refinement
+public enum Not<R: Refinement>: Refinement {
+    public typealias RefinedType = R.RefinedType
+    public static func isValid(_ value: R.RefinedType) -> Bool {
+        return !R.isValid(value)
+    }
+}
+
+// MARK: - OneOf
+
+/// a refinement that can be used to combine any two other refinements (`L` and `R`), such that either one or the other must pass
+public enum OneOf<L: Refinement, R: Refinement>: Refinement where L.RefinedType == R.RefinedType {
```

Comment:
Having `R` stand for "right side" here, but just be shorthand for `Refinement` elsewhere (like in `Not` above) feels a bit confusing. maybe this could be `LHS` and `RHS`, or everywhere else it could be `Rule`?

Reply:
`Refined` is now `Refined<A, Rule: Refinement>` in the latest commit!

Reply:
Nice!